server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java

```diff
@@ -197,6 +197,7 @@ public InputStream readBlob(OperationPurpose purpose, String blobName, long posi
         assert BlobContainer.assertPurposeConsistency(purpose, blobName);
         final SeekableByteChannel channel = Files.newByteChannel(path.resolve(blobName));
         if (position > 0L) {
+            assert position < channel.size() : "reading from " + position + " exceeds file length " + channel.size();
```

Comment:
I think it is enough for now, I haven't checked the behavior of Azure/GCS when reads start after the blob length but if they all work like S3 and throw exceptions then we could make FS implementation also throw.